Interfaces/osuAPIV2.ts

```diff
@@ -0,0 +1,128 @@
+import axios from "axios";
+import { RateLimiter } from "limiter";
+
+export interface osuAPIV2ChatBotOptions {
+    disableRateLimiting: boolean;
+    requestsPerMinute: number;
+    baseURL: string;
+}
+
+export interface osuAPIV2ChatBotToken {
+    token: string;
+    expiresAt: Date;
+}
+
+export class osuAPIV2 {
+    private readonly clientID: string;
+    private readonly clientSecret: string;
+
+    private readonly disableRateLimiting: boolean;
+    private readonly requestsPerMinute: number;
+    private readonly baseURL: string;
+
+    private bucket?: RateLimiter;
+
+    private chatBotToken?: osuAPIV2ChatBotToken;
+
+    constructor (clientID: string, clientSecret: string, options?: osuAPIV2ChatBotOptions) {
+        this.clientID = clientID;
+        this.clientSecret = clientSecret;
+        this.disableRateLimiting = options?.disableRateLimiting || false;
+        this.requestsPerMinute = options?.requestsPerMinute || 60;
+        this.baseURL = options?.baseURL || "https://osu.ppy.sh/api/v2";
+        
+        if (!this.disableRateLimiting)
+            this.bucket = new RateLimiter({
+                tokensPerInterval: this.requestsPerMinute, 
+                interval: "minute",
+            });
+    }
+
+    public getFavouriteBeatmaps (userID: string, accessToken: string, offset?: number) {
+        let endpoint = `/users/${userID}/beatmapsets/favourite?limit=51`;
+        if (offset)
+            endpoint += `&offset=${offset}`;
+        return this.get(endpoint, accessToken);
+    }
+
+    public getPlayedBeatmaps (accessToken: string, year?: number, cursorString?: string) {
+        let endpoint = "/beatmapsets/search?played=played";
+        if (year)
+            endpoint += `&q=ranked%3D${year}`;
+        if (cursorString)
+            endpoint += `&cursor_string=${cursorString}`;
+        return this.get(endpoint, accessToken);
+    }
+
+    public getUserInfo (accessToken: string) {
+        return this.get("/me", accessToken);
+    }
+
+    public getUserFriends (accessToken: string) {
+        return this.get("/friends", accessToken);
+    }
+
+    public async sendMessage (userID: string, message: string): Promise<boolean> {
+        try {
+            const token = await this.getchatBotToken();
+            await this.post("/chat/new", {
+                target_id: userID,
+                message,
+                is_action: false,
+            }, token);
+        } catch (e) {
+            if (e) return false;
+        }
+        return true;
+    }
+
+    private async getchatBotToken (): Promise<string> {
+        if (this.chatBotToken && this.chatBotToken.expiresAt > new Date())
```

Comment:
should also consider a token expired if it expires in <300s, to prevent clock differences (potential time discrepancies in between our infra and osu infra)